### ISSUE TITLE
test_data_nft_factory.py::test_nonexistent_template_index fails in CI

### ISSUE THREAD

**Issue by trentmc** [oceanprotocol/ocean.py#1299]
I just [merged](https://github.com/oceanprotocol/ocean.py/commit/5a8e3433b5477bd93bdc18edae9d19b9104065ee) a [PR](https://github.com/oceanprotocol/ocean.py/pull/1261), that had been passing CI tests while in the PR.

But when merged, CI failed a test, that's unrelated to the PR. From [the CI log](https://github.com/oceanprotocol/ocean.py/actions/runs/4056125758/jobs/6980199337), `test_data_nft_factory.py::test_nonexistent_template_index` failed.

Here's the traceback:

```text
=================================== FAILURES ===================================
_______________________ test_nonexistent_template_index ________________________

data_nft_factory = <ocean_lib.models.data_nft_factory.DataNFTFactoryContract object at 0x7fa2ec9d7a30>
publisher_wallet = <LocalAccount '0x02354A1F160A3fd7ac8b02ee91F04104440B28E7'>

    @pytest.mark.unit
    def test_nonexistent_template_index(data_nft_factory, publisher_wallet):
        """Test erc[72](https://github.com/oceanprotocol/ocean.py/actions/runs/4056125758/jobs/6980199337#step:7:73)1 non existent template creation fail"""
        non_existent_nft_template = get_non_existent_nft_template(
            data_nft_factory, check_first=10
        )
        assert non_existent_nft_template >= 0, "Non existent NFT template not found."
    
        with pytest.raises(Exception, match="Missing NFTCreated event"):
>           data_nft_factory.create(
                DataNFTArguments(
                    "DT1", "DTSYMBOL", template_index=non_existent_nft_template
                ),
                {"from": publisher_wallet, "required_confs": 0},
            )

ocean_lib/models/test/test_data_nft_factory.py:334: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ocean_lib.models.data_nft_factory.DataNFTFactoryContract object at 0x7fa2ec9d7a30>
data_nft_args = <ocean_lib.models.data_nft.DataNFTArguments object at 0x7fa2e90c24f0>
tx_dict = {'from': <LocalAccount '0x02354A1F160A3fd7ac8b02ee91F04104440B28E7'>, 'required_confs': 0}

    def create(self, data_nft_args, tx_dict):
>       return data_nft_args.deploy_contract(self.config_dict, tx_dict)

ocean_lib/models/data_nft_factory.py:34: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ocean_lib.models.data_nft.DataNFTArguments object at 0x7fa2e90c24f0>
config_dict = {'ADDRESS_FILE': '/home/runner/.ocean/ocean-contracts/artifacts/address.json', 'DOWNLOADS_PATH': 'consume-downloads', 'METADATA_CACHE_URI': 'http://172.15.0.5:5000', 'NETWORK_NAME': 'development', ...}
tx_dict = {'from': <LocalAccount '0x02354A1F160A3fd7ac8b02ee91F04104440B28E7'>, 'required_confs': 0}

    def deploy_contract(self, config_dict, tx_dict) -> DataNFT:
        from ocean_lib.models.data_nft_factory import (  # isort:skip
            DataNFTFactoryContract,
        )
    
        address = get_address_of_type(config_dict, DataNFTFactoryContract.CONTRACT_NAME)
        data_nft_factory = DataNFTFactoryContract(config_dict, address)
    
        wallet_address = get_from_address(tx_dict)
    
>       receipt = data_nft_factory.deployERC721Contract(
            self.name,
            self.symbol,
            self.template_index,
            self.additional_metadata_updater,
            self.additional_datatoken_deployer,
            self.uri,
            self.transferable,
            self.owner or wallet_address,
            tx_dict,
        )

ocean_lib/models/data_nft.py:131: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ContractTx 'deployERC721Contract(string name, string symbol, uint256 _templateIn...nalMetaDataUpdater, string tokenURI, bool transferable, address owner)'>
args = ('DT1', 'DTSYMBOL', 0, '0x0000000000000000000000000000000000000000', '0x0000000000000000000000000000000000000000', b'd...NSAyMCw3MiAzMSw3MUM0MSw2OSA1Myw2OSA2NSw3MEM3Niw3MCA4Nyw3MiA5OSw3NEw5OSw5OVonJTNFJTNDL3BhdGglM0UlM0Mvc3ZnJTNFIn0=', ...)

    def __call__(self, *args: Tuple) -> TransactionReceiptType:
        """
        Broadcast a transaction that calls this contract method.
    
        Arguments
        ---------
        *args
            Contract method inputs. You can optionally provide a
            dictionary of transaction properties as the last arg.
    
        Returns
        -------
        TransactionReceipt
            Object representing the broadcasted transaction.
        """
    
>       return self.transact(*args)

/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/brownie/network/contract.py:1864: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ContractTx 'deployERC721Contract(string name, string symbol, uint256 _templateIn...nalMetaDataUpdater, string tokenURI, bool transferable, address owner)'>
args = ('DT1', 'DTSYMBOL', 0, '0x0000000000000000000000000000000000000000', '0x0000000000000000000000000000000000000000', b'd...NSAyMCw3MiAzMSw3MUM0MSw2OSA1Myw2OSA2NSw3MEM3Niw3MCA4Nyw3MiA5OSw3NEw5OSw5OVonJTNFJTNDL3BhdGglM0UlM0Mvc3ZnJTNFIn0=', ...)
tx = {'allow_revert': None, 'from': <LocalAccount '0x02354A1F160A3fd7ac8b02ee91F04104440B28E7'>, 'gas': None, 'gas_buffer': None, ...}

    def transact(self, *args: Tuple) -> TransactionReceiptType:
        """
        Broadcast a transaction that calls this contract method.
    
        Arguments
        ---------
        *args
            Contract method inputs. You can optionally provide a
            dictionary of transaction properties as the last arg.
    
        Returns
        -------
        TransactionReceipt
            Object representing the broadcasted transaction.
        """
    
        args, tx = _get_tx(self._owner, args)
        if not tx["from"]:
            raise AttributeError(
                "Final argument must be a dict of transaction parameters that "
                "includes a `from` field specifying the sender of the transaction"
            )
    
>       return tx["from"].transfer(
            self._address,
            tx["value"],
            gas_limit=tx["gas"],
            gas_buffer=tx.get("gas_buffer"),
            gas_price=tx.get("gas_price"),
            max_fee=tx.get("max_fee"),
            priority_fee=tx.get("priority_fee"),
            nonce=tx["nonce"],
            required_confs=tx["required_confs"],
            data=self.encode_input(*args),
            allow_revert=tx["allow_revert"],
        )

/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/brownie/network/contract.py:1[73](https://github.com/oceanprotocol/ocean.py/actions/runs/4056125758/jobs/6980199337#step:7:74)7: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <LocalAccount '0x02354A1F160A3fd7ac8b02ee91F04104440B28E7'>
to = '0x8930D1F61Defb9CBD652136B4C3D93f19AC35678', amount = 0, gas_limit = None
gas_buffer = None, gas_price = None, max_fee = None, priority_fee = None
data = '0x3147c79800000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000...4f5377334e4577354f5377354f566f6e4a544e464a544e444c3342686447676c4d30556c4d304d7663335a6e4a544e46496e303d00000000000000'
nonce = None, required_confs = 0, allow_revert = None, silent = None

    def transfer(
        self,
        to: "Account" = None,
        amount: int = 0,
        gas_limit: Optional[int] = None,
        gas_buffer: Optional[float] = None,
        gas_price: Optional[int] = None,
        max_fee: Optional[int] = None,
        priority_fee: Optional[int] = None,
        data: str = None,
        nonce: Optional[int] = None,
        required_confs: int = 1,
        allow_revert: bool = None,
        silent: bool = None,
    ) -> TransactionReceipt:
        """
        Broadcast a transaction from this account.
    
        Kwargs:
            to: Account instance or address string to transfer to.
            amount: Amount of ether to send, in wei.
            gas_limit: Gas limit of the transaction.
            gas_buffer: Multiplier to apply to gas limit.
            gas_price: Gas price of legacy transaction.
            max_fee: Max fee per gas of dynamic fee transaction.
            priority_fee: Max priority fee per gas of dynamic fee transaction.
            nonce: Nonce to use for the transaction.
            data: Hexstring of data to include in transaction.
            silent: Toggles console verbosity.
    
        Returns:
            TransactionReceipt object
        """
    
        receipt, exc = self._make_transaction(
            to,
            amount,
            gas_limit,
            gas_buffer,
            gas_price,
            max_fee,
            priority_fee,
            data or "",
            nonce,
            "",
            required_confs,
            allow_revert,
            silent,
        )
    
        if rpc.is_active():
            undo_thread = threading.Thread(
                target=Chain()._add_to_undo_buffer,
                args=(
                    receipt,
                    self.transfer,
                    [],
                    {
                        "to": to,
                        "amount": amount,
                        "gas_limit": gas_limit,
                        "gas_buffer": gas_buffer,
                        "gas_price": gas_price,
                        "max_fee": max_fee,
                        "priority_fee": priority_fee,
                        "data": data,
                    },
                ),
                daemon=True,
            )
            undo_thread.start()
    
>       receipt._raise_if_reverted(exc)

/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/brownie/network/account.py:682: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Transaction '0x1e80951c6d762f76c3f1e8e0b1dfc8d12125861f1f8732ba3aa8ab8e544af274'>
exc = VirtualMachineError(ValueError({'message': 'VM Exception while processing transaction: revert None', 'code': -32000, '...c8d12125861f1f8732ba3aa8ab8e544af274': {'error': 'revert', 'program_counter': None, 'return': None, 'reason': None}}}))

    def _raise_if_reverted(self, exc: Any) -> None:
        if self.status or CONFIG.mode == "console":
            return
        if not web3.supports_traces:
            # if traces are not available, do not attempt to determine the revert reason
            raise exc or ValueError("Execution reverted")
    
        if self._dev_revert_msg is None:
            # no revert message and unable to check dev string - have to get trace
            self._expand_trace()
        if self.contract_address:
            source = ""
        elif CONFIG.argv["revert"]:
            source = self._traceback_string()
        else:
            source = self._error_string(1)
            contract = state._find_contract(self.receiver)
            if contract:
                marker = "//" if contract._build["language"] == "Solidity" else "#"
                line = self._traceback_string().split("\n")[-1]
                if marker + " dev: " in line:
                    self._dev_revert_msg = line[line.index(marker) + len(marker) : -5].strip()
    
>       raise exc._with_attr(
            source=source, revert_msg=self._revert_msg, dev_revert_msg=self._dev_revert_msg
        )
E       brownie.exceptions.VirtualMachineError: revert: ERC721DTFactory: Template index doesnt exist

/opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/brownie/network/transaction.py:446: VirtualMachineError

During handling of the above exception, another exception occurred:

data_nft_factory = <ocean_lib.models.data_nft_factory.DataNFTFactoryContract object at 0x7fa2ec9d7a30>
publisher_wallet = <LocalAccount '0x02354A1F160A3fd7ac8b02ee91F04104440B28E7'>

    @pytest.mark.unit
    def test_nonexistent_template_index(data_nft_factory, publisher_wallet):
        """Test erc721 non existent template creation fail"""
        non_existent_nft_template = get_non_existent_nft_template(
            data_nft_factory, check_first=10
        )
        assert non_existent_nft_template >= 0, "Non existent NFT template not found."
    
        with pytest.raises(Exception, match="Missing NFTCreated event"):
>           data_nft_factory.create(
                DataNFTArguments(
                    "DT1", "DTSYMBOL", template_index=non_existent_nft_template
                ),
                {"from": publisher_wallet, "required_confs": 0},
            )
E           AssertionError: Regex pattern 'Missing NFTCreated event' does not match 'revert: ERC721DTFactory: Template index doesnt exist'.

ocean_lib/models/test/test_data_nft_factory.py:[334](https://github.com/oceanprotocol/ocean.py/actions/runs/4056125758/jobs/6980199337#step:7:335): AssertionError
```

**Comment by calina-c** [comment]
There were some events a couple of weeks ago where we suddenly had race condition failures and we fixed them manually. In the meantime, brownie was updated and they also fixed the same thing, in turn rendering our fix broken. We reverted those changes and upgraded brownie, and we are now back to square 1, but it took some time to adjust back the code. We kept seeing such hiccups but I think we cleaned them all up in the meantime. I haven't seen any of these tests fail again recently.

TLDR: this was during a time of test fixes specifically related to "intended failure in tests". Haven't seen it since. I propose we close it and revisit in case it ever appears again.

CC @trentmc
